[query-cache] EvictionTest.testMaxSizeEvictionWorks

```
java.lang.AssertionError: Cache size is 101, failed to complete within 120 seconds, count left: 3
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at com.hazelcast.test.HazelcastTestSupport.assertOpenEventually(HazelcastTestSupport.java:877)
	at com.hazelcast.test.HazelcastTestSupport.assertOpenEventually(HazelcastTestSupport.java:863)
	at com.hazelcast.map.impl.querycache.EvictionTest.testMaxSizeEvictionWorks(EvictionTest.java:59)
```

https://hazelcast-l337.ci.cloudbees.com/view/Hazelcast/job/Hazelcast-3.x-IbmJDK1.7/com.hazelcast$hazelcast/1167/testReport/junit/com.hazelcast.map.impl.querycache/EvictionTest/testMaxSizeEvictionWorks/

```
java.lang.AssertionError: Cache size is 101, failed to complete within 120 seconds, count left: 1
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at com.hazelcast.test.HazelcastTestSupport.assertOpenEventually(HazelcastTestSupport.java:882)
	at com.hazelcast.test.HazelcastTestSupport.assertOpenEventually(HazelcastTestSupport.java:868)
	at com.hazelcast.map.impl.querycache.EvictionTest.testMaxSizeEvictionWorks(EvictionTest.java:75)
```

https://hazelcast-l337.ci.cloudbees.com/view/Hazelcast/job/Hazelcast-3.x-IbmJDK1.7/com.hazelcast$hazelcast/1226/testReport/junit/com.hazelcast.map.impl.querycache/EvictionTest/testMaxSizeEvictionWorks/

*maintenance*
```
java.lang.AssertionError: Cache size is 101, failed to complete within 120 seconds, count left: 9
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at com.hazelcast.test.HazelcastTestSupport.assertOpenEventually(HazelcastTestSupport.java:898)
	at com.hazelcast.test.HazelcastTestSupport.assertOpenEventually(HazelcastTestSupport.java:884)
	at com.hazelcast.map.impl.querycache.EvictionTest.testMaxSizeEvictionWorks(EvictionTest.java:75)
```

https://hazelcast-l337.ci.cloudbees.com/view/Hazelcast%20Maintenance/job/Hazelcast-3.maintenance-IbmJDK1.8/com.hazelcast$hazelcast/202/testReport/junit/com.hazelcast.map.impl.querycache/EvictionTest/testMaxSizeEvictionWorks/

```
java.lang.AssertionError: Cache size is 101, failed to complete within 120 seconds, count left: 3
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at com.hazelcast.test.HazelcastTestSupport.assertOpenEventually(HazelcastTestSupport.java:963)
	at com.hazelcast.test.HazelcastTestSupport.assertOpenEventually(HazelcastTestSupport.java:949)
	at com.hazelcast.map.impl.querycache.EvictionTest.testMaxSizeEvictionWorks(EvictionTest.java:75)
```

https://hazelcast-l337.ci.cloudbees.com/view/Hazelcast/job/Hazelcast-3.x-IbmJDK1.8/com.hazelcast$hazelcast/619/testReport/junit/com.hazelcast.map.impl.querycache/EvictionTest/testMaxSizeEvictionWorks/

```
java.lang.AssertionError: Cache size is 101, failed to complete within 120 seconds, count left: 2
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at com.hazelcast.test.HazelcastTestSupport.assertOpenEventually(HazelcastTestSupport.java:996)
	at com.hazelcast.test.HazelcastTestSupport.assertOpenEventually(HazelcastTestSupport.java:982)
	at com.hazelcast.map.impl.querycache.EvictionTest.testMaxSizeEvictionWorks(EvictionTest.java:75)
```

https://hazelcast-l337.ci.cloudbees.com/view/Hazelcast/job/Hazelcast-3.x-IbmJDK1.7/com.hazelcast$hazelcast/1390/testReport/junit/com.hazelcast.map.impl.querycache/EvictionTest/testMaxSizeEvictionWorks/

--- a/hazelcast/src/main/java/com/hazelcast/cache/impl/AbstractCacheRecordStore.java
+++ b/hazelcast/src/main/java/com/hazelcast/cache/impl/AbstractCacheRecordStore.java
@@ -271,16 +271,17 @@ public abstract class AbstractCacheRecordStore<R extends CacheRecord, CRM extend
     }
 
     @Override
-    public int evictIfRequired() {
-        int evictedCount = 0;
-        if (isEvictionEnabled()) {
-            evictedCount = evictionStrategy.evict(records, evictionPolicyEvaluator, evictionChecker, this);
-            if (isStatisticsEnabled() && evictedCount > 0) {
-                statistics.increaseCacheEvictions(evictedCount);
-            }
+    public boolean evictIfRequired() {
+        if (!isEvictionEnabled()) {
+            return false;
         }
-        return evictedCount;
-    }
+
+        boolean evicted = evictionStrategy.evict(records, evictionPolicyEvaluator, evictionChecker, this);
+        if (isStatisticsEnabled() && evicted) {
+            statistics.increaseCacheEvictions(1);
+        }
+        return evicted;
+}
 
     protected Data toData(Object obj) {
         if (obj instanceof Data) {

--- a/hazelcast/src/main/java/com/hazelcast/cache/impl/ICacheRecordStore.java
+++ b/hazelcast/src/main/java/com/hazelcast/cache/impl/ICacheRecordStore.java
@@ -439,9 +439,9 @@ public interface ICacheRecordStore {
      * Evict cache record store if eviction is required.
      * <p>Eviction logic is handled as specified {@link com.hazelcast.config.EvictionPolicy}
      * in {@link com.hazelcast.config.CacheConfig} for this record store</p>
-     * @return the number of evicted records.
+     * @return true is an entry was evicted, otherwise false
      */
-    int evictIfRequired();
+    boolean evictIfRequired();
 
     /**
      * Determines whether wan replication is enabled or not for this record store.

--- a/hazelcast/src/main/java/com/hazelcast/cache/impl/record/CacheRecordHashMap.java
+++ b/hazelcast/src/main/java/com/hazelcast/cache/impl/record/CacheRecordHashMap.java
@@ -196,7 +196,7 @@ public class CacheRecordHashMap
 
     @Override
     public <C extends EvictionCandidate<Data, CacheRecord>> boolean tryEvict(C evictionCandidate,
-                                                                             EvictionListener<Data, CacheRecord> evictionListener) {
+                                                                 EvictionListener<Data, CacheRecord> evictionListener) {
         if (evictionCandidate == null) {
             return false;
         }

--- a/hazelcast/src/main/java/com/hazelcast/cache/impl/record/CacheRecordHashMap.java
+++ b/hazelcast/src/main/java/com/hazelcast/cache/impl/record/CacheRecordHashMap.java
@@ -195,21 +195,18 @@ public class CacheRecordHashMap
     }
 
     @Override
-    public <C extends EvictionCandidate<Data, CacheRecord>> int evict(Iterable<C> evictionCandidates,
-                                                                      EvictionListener<Data, CacheRecord> evictionListener) {
-        if (evictionCandidates == null) {
-            return 0;
+    public <C extends EvictionCandidate<Data, CacheRecord>> boolean evict(C evictionCandidate,
+                                                                          EvictionListener<Data, CacheRecord> evictionListener) {
+        if (evictionCandidate == null) {
+            return false;
         }
-        int actualEvictedCount = 0;
-        for (EvictionCandidate<Data, CacheRecord> evictionCandidate : evictionCandidates) {
-            if (remove(evictionCandidate.getAccessor()) != null) {
-                actualEvictedCount++;
-                if (evictionListener != null) {
-                    evictionListener.onEvict(evictionCandidate.getAccessor(), evictionCandidate.getEvictable(), false);
-                }
-            }
+        if (remove(evictionCandidate.getAccessor()) == null) {
+            return false;
         }
-        return actualEvictedCount;
+        if (evictionListener != null) {
+            evictionListener.onEvict(evictionCandidate.getAccessor(), evictionCandidate.getEvictable(), false);
+        }
+        return true;
     }
 
     @Override

--- a/hazelcast/src/main/java/com/hazelcast/cache/impl/record/CacheRecordHashMap.java
+++ b/hazelcast/src/main/java/com/hazelcast/cache/impl/record/CacheRecordHashMap.java
@@ -195,8 +195,8 @@ public class CacheRecordHashMap
     }
 
     @Override
-    public <C extends EvictionCandidate<Data, CacheRecord>> boolean evict(C evictionCandidate,
-                                                                          EvictionListener<Data, CacheRecord> evictionListener) {
+    public <C extends EvictionCandidate<Data, CacheRecord>> boolean tryEvict(C evictionCandidate,
+                                                                             EvictionListener<Data, CacheRecord> evictionListener) {
         if (evictionCandidate == null) {
             return false;
         }

--- a/hazelcast/src/main/java/com/hazelcast/cache/impl/record/SampleableCacheRecordMap.java
+++ b/hazelcast/src/main/java/com/hazelcast/cache/impl/record/SampleableCacheRecordMap.java
@@ -19,7 +19,7 @@ package com.hazelcast.cache.impl.record;
 import com.hazelcast.internal.eviction.impl.strategy.sampling.SampleableEvictableStore;
 import com.hazelcast.nio.serialization.Data;
 
-public interface SampleableCacheRecordMap<K extends Data, V extends CacheRecord>
+public interface  SampleableCacheRecordMap<K extends Data, V extends CacheRecord>
         extends CacheRecordMap<K, V>, SampleableEvictableStore<K, V> {
 
 }

--- a/hazelcast/src/main/java/com/hazelcast/internal/eviction/EvictableStore.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/eviction/EvictableStore.java
@@ -37,7 +37,7 @@ public interface EvictableStore<A, E extends Evictable> {
      *
      * @return <code>true</code> if the candidate was evicted, otherwise returns <code>false</code>
      */
-    <C extends EvictionCandidate<A, E>> boolean evict(C evictionCandidate,
-                                                  EvictionListener<A, E> evictionListener);
+    <C extends EvictionCandidate<A, E>> boolean tryEvict(C evictionCandidate,
+                                                         EvictionListener<A, E> evictionListener);
 
 }

--- a/hazelcast/src/main/java/com/hazelcast/internal/eviction/EvictableStore.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/eviction/EvictableStore.java
@@ -27,14 +27,17 @@ public interface EvictableStore<A, E extends Evictable> {
 
     /**
      * The evict method is called by the {@link EvictionStrategy} to eventually evict, by the policy, selected
-     * candidates from the internal data structures.
+     * candidate from the internal data structures.
      *
-     * @param evictionCandidates Multiple {@link EvictionCandidate} to be evicted
+     * It could be the selected evicted candidate cannot be evicted. For example it might be gone at the time when
+     * the method is executed. In this case it will indicate this condition by returning false.
+     *
+     * @param evictionCandidate {@link EvictionCandidate} to be evicted
      * @param evictionListener   {@link EvictionListener} to listen evicted entries
      *
-     * @return evicted entry count
+     * @return <code>true</code> if the candidate was evicted, otherwise returns <code>false</code>
      */
-    <C extends EvictionCandidate<A, E>> int evict(Iterable<C> evictionCandidates,
+    <C extends EvictionCandidate<A, E>> boolean evict(C evictionCandidate,
                                                   EvictionListener<A, E> evictionListener);
 
 }

--- a/hazelcast/src/main/java/com/hazelcast/internal/eviction/impl/evaluator/EvictionPolicyEvaluator.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/eviction/impl/evaluator/EvictionPolicyEvaluator.java
@@ -50,7 +50,7 @@ public class EvictionPolicyEvaluator<A, E extends Evictable> {
      * @param evictionCandidates Multiple {@link com.hazelcast.internal.eviction.EvictionCandidate} to be evicted
      * @return multiple {@link com.hazelcast.internal.eviction.EvictionCandidate} these are available to be evicted
      */
-    public <C extends EvictionCandidate<A, E>> Iterable<C> evaluate(Iterable<C> evictionCandidates) {
+    public <C extends EvictionCandidate<A, E>> C evaluate(Iterable<C> evictionCandidates) {
         C selectedEvictionCandidate = null;
         long now = Clock.currentTimeMillis();
         for (C currentEvictionCandidate : evictionCandidates) {
@@ -59,7 +59,7 @@ public class EvictionPolicyEvaluator<A, E extends Evictable> {
             } else {
                 E evictable = currentEvictionCandidate.getEvictable();
                 if (isExpired(now, evictable)) {
-                    return returnEvictionCandidate(currentEvictionCandidate);
+                    return currentEvictionCandidate;
                 }
 
                 int comparisonResult = evictionPolicyComparator.compare(selectedEvictionCandidate, currentEvictionCandidate);
@@ -68,15 +68,7 @@ public class EvictionPolicyEvaluator<A, E extends Evictable> {
                 }
             }
         }
-        return returnEvictionCandidate(selectedEvictionCandidate);
-    }
-
-    private <C extends EvictionCandidate<A, E>> Iterable<C> returnEvictionCandidate(C evictionCandidate) {
-        if (evictionCandidate == null) {
-            return null;
-        } else {
-            return evictionCandidate instanceof Iterable ? (Iterable<C>) evictionCandidate : singleton(evictionCandidate);
-        }
+        return selectedEvictionCandidate;
     }
 
     private boolean isExpired(long now, Evictable evictable) {

--- a/hazelcast/src/main/java/com/hazelcast/internal/eviction/impl/evaluator/EvictionPolicyEvaluator.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/eviction/impl/evaluator/EvictionPolicyEvaluator.java
@@ -42,11 +42,13 @@ public class EvictionPolicyEvaluator<A, E extends Evictable> {
     }
 
     /**
-     * The evaluate method implements the {@link com.hazelcast.config.EvictionPolicy} rule
-     * on the given input set of candidates.
+     * Selects the best candidate to be evicted.
+     * The definition of the best depends on configured eviction policy. (LRU, LFU, custom, etc)
+     *
+     * It returns <code>null</code> when there the input is empty.
      *
      * @param evictionCandidates Multiple {@link com.hazelcast.internal.eviction.EvictionCandidate} to be evicted
-     * @return multiple {@link com.hazelcast.internal.eviction.EvictionCandidate} these are available to be evicted
+     * @return a selected candidate to be evicted or null.
      */
     public <C extends EvictionCandidate<A, E>> C evaluate(Iterable<C> evictionCandidates) {
         C selectedEvictionCandidate = null;

--- a/hazelcast/src/main/java/com/hazelcast/internal/eviction/impl/evaluator/EvictionPolicyEvaluator.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/eviction/impl/evaluator/EvictionPolicyEvaluator.java
@@ -22,8 +22,6 @@ import com.hazelcast.internal.eviction.EvictionPolicyComparator;
 import com.hazelcast.internal.eviction.Expirable;
 import com.hazelcast.util.Clock;
 
-import static java.util.Collections.singleton;
-
 /**
  * Default {@link EvictionPolicyEvaluator} implementation.
  *

--- a/hazelcast/src/main/java/com/hazelcast/internal/eviction/impl/strategy/sampling/SamplingEvictionStrategy.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/eviction/impl/strategy/sampling/SamplingEvictionStrategy.java
@@ -45,15 +45,15 @@ public final class SamplingEvictionStrategy<A, E extends Evictable, S extends Sa
      *                                  eviction is required or not.
      * @param evictionListener          {@link EvictionListener} to listen evicted entries
      *
-     * @return evicted entry count
+     * @return true is an entry was evicted, otherwise false
      */
-    public int evict(S evictableStore, EvictionPolicyEvaluator<A, E> evictionPolicyEvaluator,
+    public boolean evict(S evictableStore, EvictionPolicyEvaluator<A, E> evictionPolicyEvaluator,
                      EvictionChecker evictionChecker, EvictionListener<A, E> evictionListener) {
         if (evictionChecker != null) {
             if (evictionChecker.isEvictionRequired()) {
                 return evictInternal(evictableStore, evictionPolicyEvaluator, evictionListener);
             } else {
-                return 0;
+                return false;
             }
         } else {
             return evictInternal(evictableStore, evictionPolicyEvaluator, evictionListener);
@@ -67,14 +67,14 @@ public final class SamplingEvictionStrategy<A, E extends Evictable, S extends Sa
      * @param evictionPolicyEvaluator   {@link EvictionPolicyEvaluator} to evaluate
      * @param evictionListener          {@link EvictionListener} to listen evicted entries
      *
-     * @return evicted entry count
+     * @return true is an entry was evicted, otherwise false
      */
-    protected int evictInternal(S sampleableEvictableStore,
+    protected boolean evictInternal(S sampleableEvictableStore,
             EvictionPolicyEvaluator<A, E> evictionPolicyEvaluator,
             EvictionListener<A, E> evictionListener) {
         final Iterable<EvictionCandidate<A, E>> samples = sampleableEvictableStore.sample(SAMPLE_COUNT);
-        final Iterable<EvictionCandidate<A, E>> evictionCandidates = evictionPolicyEvaluator.evaluate(samples);
-        return sampleableEvictableStore.evict(evictionCandidates, evictionListener);
+        final EvictionCandidate<A, E> evictionCandidate = evictionPolicyEvaluator.evaluate(samples);
+        return sampleableEvictableStore.evict(evictionCandidate, evictionListener);
     }
 
 }

--- a/hazelcast/src/main/java/com/hazelcast/internal/eviction/impl/strategy/sampling/SamplingEvictionStrategy.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/eviction/impl/strategy/sampling/SamplingEvictionStrategy.java
@@ -74,7 +74,7 @@ public final class SamplingEvictionStrategy<A, E extends Evictable, S extends Sa
             EvictionListener<A, E> evictionListener) {
         final Iterable<EvictionCandidate<A, E>> samples = sampleableEvictableStore.sample(SAMPLE_COUNT);
         final EvictionCandidate<A, E> evictionCandidate = evictionPolicyEvaluator.evaluate(samples);
-        return sampleableEvictableStore.evict(evictionCandidate, evictionListener);
+        return sampleableEvictableStore.tryEvict(evictionCandidate, evictionListener);
     }
 
 }

--- a/hazelcast/src/main/java/com/hazelcast/internal/nearcache/impl/store/HeapNearCacheRecordMap.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/nearcache/impl/store/HeapNearCacheRecordMap.java
@@ -91,8 +91,8 @@ public class HeapNearCacheRecordMap<K, V extends NearCacheRecord>
     }
 
     @Override
-    public <C extends EvictionCandidate<K, V>> boolean evict(C evictionCandidate,
-                                                             EvictionListener<K, V> evictionListener) {
+    public <C extends EvictionCandidate<K, V>> boolean tryEvict(C evictionCandidate,
+                                                                EvictionListener<K, V> evictionListener) {
         if (evictionCandidate == null) {
             return false;
         }

--- a/hazelcast/src/main/java/com/hazelcast/internal/nearcache/impl/store/HeapNearCacheRecordMap.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/nearcache/impl/store/HeapNearCacheRecordMap.java
@@ -91,21 +91,18 @@ public class HeapNearCacheRecordMap<K, V extends NearCacheRecord>
     }
 
     @Override
-    public <C extends EvictionCandidate<K, V>> int evict(Iterable<C> evictionCandidates,
-                                                         EvictionListener<K, V> evictionListener) {
-        if (evictionCandidates == null) {
-            return 0;
+    public <C extends EvictionCandidate<K, V>> boolean evict(C evictionCandidate,
+                                                             EvictionListener<K, V> evictionListener) {
+        if (evictionCandidate == null) {
+            return false;
         }
-        int actualEvictedCount = 0;
-        for (EvictionCandidate<K, V> evictionCandidate : evictionCandidates) {
-            if (remove(evictionCandidate.getAccessor()) != null) {
-                actualEvictedCount++;
-                if (evictionListener != null) {
-                    evictionListener.onEvict(evictionCandidate.getAccessor(), evictionCandidate.getEvictable(), false);
-                }
-            }
+        if (remove(evictionCandidate.getAccessor()) == null) {
+            return false;
         }
-        return actualEvictedCount;
+        if (evictionListener != null) {
+            evictionListener.onEvict(evictionCandidate.getAccessor(), evictionCandidate.getEvictable(), false);
+        }
+        return true;
     }
 
     @Override

--- a/hazelcast/src/main/java/com/hazelcast/map/impl/querycache/subscriber/EvictionOperator.java
+++ b/hazelcast/src/main/java/com/hazelcast/map/impl/querycache/subscriber/EvictionOperator.java
@@ -58,12 +58,11 @@ public class EvictionOperator {
         return evictionStrategy != null && evictionPolicyEvaluator != null;
     }
 
-    int evictIfRequired() {
-        int evictedCount = 0;
+    boolean evictIfRequired() {
         if (isEvictionEnabled()) {
-            evictedCount = evictionStrategy.evict(cache, evictionPolicyEvaluator, evictionChecker, listener);
+            return evictionStrategy.evict(cache, evictionPolicyEvaluator, evictionChecker, listener);
         }
-        return evictedCount;
+        return false;
     }
 
     private EvictionChecker createCacheEvictionChecker() {

--- a/hazelcast/src/main/java/com/hazelcast/map/impl/querycache/subscriber/EvictionOperator.java
+++ b/hazelcast/src/main/java/com/hazelcast/map/impl/querycache/subscriber/EvictionOperator.java
@@ -26,12 +26,14 @@ import com.hazelcast.map.impl.querycache.subscriber.record.QueryCacheRecord;
 import com.hazelcast.nio.serialization.Data;
 
 import static com.hazelcast.internal.config.ConfigValidator.checkEvictionConfig;
+import static com.hazelcast.internal.eviction.EvictionChecker.EVICT_ALWAYS;
 import static com.hazelcast.internal.eviction.EvictionPolicyEvaluatorProvider.getEvictionPolicyEvaluator;
 
 /**
  * Contains eviction specific functionality of a {@link QueryCacheRecordStore}.
  */
 public class EvictionOperator {
+    private static final int MAX_EVICTION_ATTEMPTS = 10;
 
     private final QueryCacheRecordHashMap cache;
     private final EvictionConfig evictionConfig;
@@ -58,18 +60,22 @@ public class EvictionOperator {
         return evictionStrategy != null && evictionPolicyEvaluator != null;
     }
 
-    boolean evictIfRequired() {
-        if (isEvictionEnabled()) {
-            return evictionStrategy.evict(cache, evictionPolicyEvaluator, evictionChecker, listener);
+    void evictIfRequired() {
+        if (!isEvictionEnabled()) {
+            return;
         }
-        return false;
+
+        for (int i = 0; evictionChecker.isEvictionRequired() && i < MAX_EVICTION_ATTEMPTS; i++) {
+            // we already established we should evict -> we can pass EVICT_ALWAYS to the eviction strategy
+            evictionStrategy.evict(cache, evictionPolicyEvaluator, EVICT_ALWAYS, listener);
+        }
     }
 
     private EvictionChecker createCacheEvictionChecker() {
         return new EvictionChecker() {
             @Override
             public boolean isEvictionRequired() {
-                return cache.size() > evictionConfig.getSize();
+                return cache.size() >= evictionConfig.getSize();
             }
         };
     }

--- a/hazelcast/src/main/java/com/hazelcast/map/impl/querycache/subscriber/EvictionOperator.java
+++ b/hazelcast/src/main/java/com/hazelcast/map/impl/querycache/subscriber/EvictionOperator.java
@@ -33,6 +33,11 @@ import static com.hazelcast.internal.eviction.EvictionPolicyEvaluatorProvider.ge
  * Contains eviction specific functionality of a {@link QueryCacheRecordStore}.
  */
 public class EvictionOperator {
+    // It could be the current size of the CQC is over a configured limit. This can happen e.g. when multiple threads
+    // are inserting entries concurrently. However each eviction cycle can remove at most 1 entry -> we run multiple
+    // eviction cycles when the current size is over the configured eviction threshold. This property controls maximum
+    // no. of eviction cycles during one insertion into CQC.
+    // Too low value might be insufficient to properly evict entries, too high value can cause latency spikes.
     private static final int MAX_EVICTION_ATTEMPTS = 10;
 
     private final QueryCacheRecordHashMap cache;

--- a/hazelcast/src/main/java/com/hazelcast/map/impl/querycache/subscriber/QueryCacheRecordHashMap.java
+++ b/hazelcast/src/main/java/com/hazelcast/map/impl/querycache/subscriber/QueryCacheRecordHashMap.java
@@ -110,7 +110,7 @@ public class QueryCacheRecordHashMap extends SampleableConcurrentHashMap<Data, Q
 
     @Override
     public <C extends EvictionCandidate<Data, QueryCacheRecord>>
-    boolean evict(C evictionCandidate, EvictionListener<Data, QueryCacheRecord> evictionListener) {
+    boolean tryEvict(C evictionCandidate, EvictionListener<Data, QueryCacheRecord> evictionListener) {
         if (evictionCandidate == null) {
             return false;
         }

--- a/hazelcast/src/main/java/com/hazelcast/map/impl/querycache/subscriber/QueryCacheRecordHashMap.java
+++ b/hazelcast/src/main/java/com/hazelcast/map/impl/querycache/subscriber/QueryCacheRecordHashMap.java
@@ -110,20 +110,17 @@ public class QueryCacheRecordHashMap extends SampleableConcurrentHashMap<Data, Q
 
     @Override
     public <C extends EvictionCandidate<Data, QueryCacheRecord>>
-    int evict(Iterable<C> evictionCandidates, EvictionListener<Data, QueryCacheRecord> evictionListener) {
-        if (evictionCandidates == null) {
-            return 0;
+    boolean evict(C evictionCandidate, EvictionListener<Data, QueryCacheRecord> evictionListener) {
+        if (evictionCandidate == null) {
+            return false;
         }
-        int actualEvictedCount = 0;
-        for (EvictionCandidate<Data, QueryCacheRecord> evictionCandidate : evictionCandidates) {
-            if (remove(evictionCandidate.getAccessor()) != null) {
-                actualEvictedCount++;
-                if (evictionListener != null) {
-                    evictionListener.onEvict(evictionCandidate.getAccessor(), evictionCandidate.getEvictable(), false);
-                }
-            }
+        if (remove(evictionCandidate.getAccessor()) == null) {
+            return false;
         }
-        return actualEvictedCount;
+        if (evictionListener != null) {
+            evictionListener.onEvict(evictionCandidate.getAccessor(), evictionCandidate.getEvictable(), false);
+        }
+        return true;
     }
 
     @Override

--- a/hazelcast/src/test/java/com/hazelcast/internal/eviction/EvictionPolicyEvaluatorTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/internal/eviction/EvictionPolicyEvaluatorTest.java
@@ -150,19 +150,10 @@ public class EvictionPolicyEvaluatorTest extends HazelcastTestSupport {
 
         sleepAtLeastMillis(1);
 
-        Iterable<EvictionCandidate<Integer, CacheObjectRecord>> evictedRecords =
-                evictionPolicyEvaluator.evaluate(records);
+        EvictionCandidate<Integer, CacheObjectRecord> evictionCandidate = evictionPolicyEvaluator.evaluate(records);
+        assertNotNull(evictionCandidate);
 
-        assertNotNull(evictedRecords);
-
-        Iterator<EvictionCandidate<Integer, CacheObjectRecord>> evictedRecordsIterator = evictedRecords.iterator();
-        assertTrue(evictedRecordsIterator.hasNext());
-
-        EvictionCandidate<Integer, CacheObjectRecord> candidateEvictedRecord = evictedRecordsIterator.next();
-        assertNotNull(candidateEvictedRecord);
-        assertFalse(evictedRecordsIterator.hasNext());
-
-        CacheObjectRecord evictedRecord = candidateEvictedRecord.getEvictable();
+        CacheObjectRecord evictedRecord = evictionCandidate.getEvictable();
         assertNotNull(evictedRecord);
         if (useExpiredEntry) {
             assertEquals(EXPECTED_EXPIRED_RECORD_VALUE, evictedRecord.getValue());
@@ -231,19 +222,11 @@ public class EvictionPolicyEvaluatorTest extends HazelcastTestSupport {
             records.add(new SimpleEvictionCandidate<Integer, CacheObjectRecord>(i, record));
         }
 
-        Iterable<EvictionCandidate<Integer, CacheObjectRecord>> evictedRecords =
-                evictionPolicyEvaluator.evaluate(records);
+        EvictionCandidate<Integer, CacheObjectRecord> evictionCandidate = evictionPolicyEvaluator.evaluate(records);
 
-        assertNotNull(evictedRecords);
+        assertNotNull(evictionCandidate);
 
-        Iterator<EvictionCandidate<Integer, CacheObjectRecord>> evictedRecordsIterator = evictedRecords.iterator();
-        assertTrue(evictedRecordsIterator.hasNext());
-
-        EvictionCandidate<Integer, CacheObjectRecord> candidateEvictedRecord = evictedRecordsIterator.next();
-        assertNotNull(candidateEvictedRecord);
-        assertFalse(evictedRecordsIterator.hasNext());
-
-        CacheObjectRecord evictedRecord = candidateEvictedRecord.getEvictable();
+        CacheObjectRecord evictedRecord = evictionCandidate.getEvictable();
         assertNotNull(evictedRecord);
         if (useExpiredEntry) {
             assertEquals(EXPECTED_EXPIRED_RECORD_VALUE, evictedRecord.getValue());

--- a/hazelcast/src/test/java/com/hazelcast/internal/eviction/EvictionStrategyTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/internal/eviction/EvictionStrategyTest.java
@@ -108,7 +108,6 @@ public class EvictionStrategyTest<K, V extends Evictable, S extends SampleableEv
     @Test
     public void evictionPolicySuccessfullyEvaluatedOnSamplingBasedEvictionStrategy() {
         final int RECORD_COUNT = 100;
-        final int EXPECTED_EVICTED_COUNT = 1;
         final int EXPECTED_EVICTED_RECORD_VALUE = RECORD_COUNT / 2;
 
         Node node = TestUtil.getNode(instance);
@@ -140,16 +139,16 @@ public class EvictionStrategyTest<K, V extends Evictable, S extends SampleableEv
         // we are testing "EvictionStrategy" in this test, so we mock "EvictionPolicyEvaluator" (it's tested in another test)
         EvictionPolicyEvaluator evictionPolicyEvaluator = mock(EvictionPolicyEvaluator.class);
         when(evictionPolicyEvaluator.evaluate(Matchers.any(Iterable.class))).
-                thenReturn(Collections.singleton(evictionCandidate));
+                thenReturn(evictionCandidate);
         when(evictionPolicyEvaluator.getEvictionPolicyComparator()).thenReturn(null);
 
         assertEquals(RECORD_COUNT, cacheRecordMap.size());
         assertTrue(cacheRecordMap.containsKey(expectedData));
         assertTrue(cacheRecordMap.containsValue(expectedEvictedRecord));
 
-        int evictedCount = evictionStrategy.evict((S) cacheRecordMap, evictionPolicyEvaluator, EVICT_ALWAYS, NO_LISTENER);
-        assertEquals(EXPECTED_EVICTED_COUNT, evictedCount);
-        assertEquals(RECORD_COUNT - EXPECTED_EVICTED_COUNT, cacheRecordMap.size());
+        boolean evicted = evictionStrategy.evict((S) cacheRecordMap, evictionPolicyEvaluator, EVICT_ALWAYS, NO_LISTENER);
+        assertTrue(evicted);
+        assertEquals(RECORD_COUNT - 1, cacheRecordMap.size());
         assertFalse(cacheRecordMap.containsKey(expectedData));
         assertFalse(cacheRecordMap.containsValue(expectedEvictedRecord));
     }

--- a/hazelcast/src/test/java/com/hazelcast/map/impl/querycache/EvictionTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/map/impl/querycache/EvictionTest.java
@@ -25,7 +25,7 @@ import com.hazelcast.core.EntryEvent;
 import com.hazelcast.core.HazelcastInstance;
 import com.hazelcast.core.IMap;
 import com.hazelcast.map.QueryCache;
-import com.hazelcast.map.listener.EntryEvictedListener;
+import com.hazelcast.map.listener.EntryAddedListener;
 import com.hazelcast.query.TruePredicate;
 import com.hazelcast.test.HazelcastParallelClassRunner;
 import com.hazelcast.test.HazelcastTestSupport;
@@ -56,15 +56,12 @@ public class EvictionTest extends HazelcastTestSupport {
         HazelcastInstance node = createHazelcastInstance(config);
         IMap<Integer, Integer> map = getMap(node, mapName);
 
-        // expecting at least populationCount - maxSize - 50 evicted entries according to max size.
-        // 50 states an error margin since eviction does not sweep precise number of entries.
-        int margin = 50;
-        final CountDownLatch evictedCount = new CountDownLatch(populationCount - maxSize - margin);
+        final CountDownLatch entryCountingLatch = new CountDownLatch(populationCount);
         QueryCache<Integer, Integer> cache = map.getQueryCache(cacheName, TruePredicate.INSTANCE, true);
-        String listener = cache.addEntryListener(new EntryEvictedListener() {
+        String listener = cache.addEntryListener(new EntryAddedListener() {
             @Override
-            public void entryEvicted(EntryEvent event) {
-                evictedCount.countDown();
+            public void entryAdded(EntryEvent event) {
+                entryCountingLatch.countDown();
             }
         }, false);
 
@@ -72,8 +69,11 @@ public class EvictionTest extends HazelcastTestSupport {
             map.put(i, i);
         }
 
-        assertOpenEventually("Cache size is " + cache.size(), evictedCount);
-        assertQueryCacheEvicted(maxSize, margin, cache);
+        assertOpenEventually("Cache size is " + cache.size(), entryCountingLatch);
+
+        // expecting at most populationCount - maxSize - 5 entries
+        // 5 states an error margin since eviction does not sweep precise number of entries.
+        assertQueryCacheEvicted(maxSize, 5, cache);
         assertTrue(cache.removeEntryListener(listener));
     }
 

